File: java/hello-world-http/src/main/java/dev/restate/sdk/examples/Greeter.java
package dev.restate.sdk.examples;

import dev.restate.sdk.RestateContext;
import dev.restate.sdk.common.CoreSerdes;
import dev.restate.sdk.common.StateKey;
import dev.restate.sdk.examples.generated.*;
import dev.restate.sdk.http.vertx.RestateHttpEndpointBuilder;

import static dev.restate.sdk.examples.generated.GreeterProto.*;

/**
 * This service greets the users.
 */
public class Greeter extends GreeterRestate.GreeterRestateImplBase {

  // Count state. The count is per Person name.
  // See https://docs.restate.dev/services/sdk/state for more details.
  private static final StateKey<Integer> COUNT = StateKey.of("count", CoreSerdes.INT);

  @Override
  public GreetResponse greet(RestateContext context, GreetRequest request) {
    // Get the count and increment it
    int count = context.get(COUNT).orElse(1);
    context.set(COUNT, count + 1);

    // Send the response back
    return GreetResponse.newBuilder()
            .setMessage("Hello " + request.getName() + " for the " + count + " time!")
            .build();
  }

  public static void main(String[] args) {
    RestateHttpEndpointBuilder.builder()
            // Register the service Greeter
            .withService(new Greeter())
            // Start the Restate Endpoint HTTP Server
            .buildAndListen();
  }
}

File: scala/hello-world-http/src/main/scala/dev/restate/sdk/examples/Greeter.scala
package dev.restate.sdk.examples

import dev.restate.sdk.RestateService
import dev.restate.sdk.common.{CoreSerdes, StateKey}
import greeter.GreeterGrpc.GreeterImplBase
import greeter.{GreetRequest, GreetResponse}
import io.grpc.stub.StreamObserver
import dev.restate.sdk.examples.Greeter.COUNT;

object Greeter{
  private val COUNT = StateKey.of[Integer]("count", CoreSerdes.INT)
}

class Greeter extends GreeterImplBase with RestateService {
  override def greet(request: GreetRequest, responseObserver: StreamObserver[GreetResponse]): Unit = {
    val ctx = restateContext

    val count = ctx.get(COUNT).orElse(Integer.valueOf(1))
    ctx.set(COUNT, Integer.valueOf(count + 1))

    responseObserver.onNext(GreetResponse.newBuilder.setMessage(s"Hello ${request.getName} for the ${count} time!").build)
    responseObserver.onCompleted()
  }
}